### PR TITLE
Dev/data structures

### PR DESCRIPTION
**Pull request by YuvalRi**
Good luck :)

### CODE REVIEW

File: datastruc_ex6.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+            all_freq[i] += 1
+        else:
+            all_freq[i] = 1
+    return all_freq
+
+
+if __name__ == "__main__":
+    print(freqDict("yuvalritsker"))
+    print(freqDict("ofekshmuel"))
+    print(freqDict("chugokujin"))
```
Comment (ofekshmu) [inline datastruc_ex6.py:19]: Great!

File: datastruc_ex7.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    all_freq = {}
+    for i in s:
+        if i in all_freq:
+            all_freq[i] += 1
+        else:
+            all_freq[i] = 1
+    return all_freq
+
+
+def freqFunc(lst: list):
```
Comment (ofekshmu) [inline datastruc_ex7.py:40]: Bad name, does not relate to the functions functionality

File: datastruc_ex7.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
+import random as rnd
+
+# ex7
+# a
+
+
+def func(n: int, random: bool, lst: list):
```
Comment (ofekshmu) [inline datastruc_ex7.py:7]: func? no documentation? you are sleeping on the couch today

File: datastruc_ex7.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        med = (med1 + med2)/2
+        return med
+    else:
+        med = new_lst[(len(new_lst))//2]
+        return med
+
+
+if __name__ == "__main__":
+    # in this case random = True
+    a = freqDict(freqFunc(func(7, True, lst=[1, 2, 3, 4, 5, 6, 7])))
```
Comment (ofekshmu) [inline datastruc_ex7.py:70]: This line is a mess, what does all of this do?

File: datastruc_ex7.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    print(a)
+    print(maxCharacter(a))
+    print(findMedian(a))
+    # in this case random = False
+    b = freqDict(freqFunc(func(5, False, lst=[1, 2, 3, 4, 5])))
+    print(b)
+    print(maxCharacter(b))
+    print(findMedian(b))
```
Comment (ofekshmu) [inline datastruc_ex7.py:78]: Too many prints. use a single one to print every thing you want

File: datastruc_ex7.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            selected_words = []
+            for i in range(len(randomlist)):
+                selected_words.append(text_list[randomlist[i]])
+            return selected_words
+        else:
+            text_list = allText.split('\n')
+            selected_words = []
+            for i in range(len(lst)):
+                selected_words.append(text_list[lst[i]])
+            return selected_words
```
Comment (ofekshmu) [inline datastruc_ex7.py:22]: A lot of code here repeats itself in this section. for example:
line 22 and line 16 are identical.
13 and 19.
hint: 14-15 and 20-21 are ALMOST identical, you can also improve this.
REDUCE code that repeats itself.

File: datastruc_ex7.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    '''
+    A function which creates a frequency dictionary from a string
+    '''
+    all_freq = {}
+    for i in s:
+        if i in all_freq:
+            all_freq[i] += 1
+        else:
+            all_freq[i] = 1
+    return all_freq
```
Comment (ofekshmu) [inline datastruc_ex7.py:37]: I saw this function so many times in this project. Do you think that each time we need to reuse a function we COPY PASTE IT? look up in google: "import a function from a different file python"

This is also REWRITING CODE  - NOT GOOD

File: datastruc_ex8.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+    '''
+    A function which creates a frequency dictionary from a string
+    '''
+    all_freq = {}
+    for i in s:
+        if i in all_freq:
+            all_freq[i] += 1
+        else:
+            all_freq[i] = 1
+    return all_freq
```
Comment (ofekshmu) [inline datastruc_ex8.py:14]: and again...

File: datastruc_ex8.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    '''
+    A function which returns True if the received string can be created from
+    the characters available in the given dictionary and False otherwise
+    '''
+    for c in s:
+        if c not in dict.keys() or dict[c] == 0:
+            return False
+        else:
+            dict[c] -= 1
+    return True
```
Comment (ofekshmu) [inline datastruc_ex8.py:27]: Love it

**Review by ofekshmu — CHANGES_REQUESTED** [review]
The overall work is great! You improved tremendously!
You still have a lot to improve, BUT remember that all of those things are stuff that a beginner would never even bother to pay attention to. YOU are on a different LEAGUE!

File: datastruc_ex1.py [hunk 10]
```diff
@@ -0,0 +1,14 @@
+# ex1
+
+def elementSum(lst: list):
+    '''
+    function that recieves a list and returns the sum of its elements
+    '''
+    lst_sum = 0
+    for i in range(len(lst)):
+        lst_sum = lst[i] + lst_sum
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex1.py]: done

File: datastruc_ex2.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+
+def smallestElement(lst):
+    '''
+    function that recieves a list of integers and returns its smallest element
+    '''
+    if len(lst) == 0:
+        print("Error. Please enter at least one element in the list")
+    else:
+        smallest = lst[0]
+        for i in lst:
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex2.py:11]: done

File: datastruc_ex3.py [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    '''
+    if len(s) > 2:
+        return True
+
+
+def firstandLast(s: str):
+    '''
+    A function which checks if both first and last characters are the same
+    '''
+    if s != "" and s[0] == s[-1]:
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex3.py]: done. I used the first implementation noted above.
Need to check if the second implementation is possible.

File: datastruc_ex3.py [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+    '''
+    counter = 0
+    for i in lst:
+        if checkLen(i) and firstandLast(i):
+            counter = counter + 1
+    print(counter)
+
+
+if __name__ == "__main__":
+    lst = ["", "aa", "Yuval", "11", "1451", "OfekO", "LocaL"]
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex3.py:33]: Yes. What is the correct result in the case above? tell me and I will tell you the received result.

File: datastruc_ex5.py [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    new_dict = {}
+    for key, value in dict.items():
+        if dict[key] > 170:
+            new_dict[key] = value
+    return new_dict
+
+
+if __name__ == "__main__":
+    dict = {'Cierra Vega': 175, 'Alden Cantrell': 180, 'Kierra Gentry': 165, 'Pierre Cox': 190}
+    print(filterValues(dict))
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex5.py:28]: thanks

File: datastruc_ex6.py (hunk 1 above)
Comment (YuvalRi) [reply datastruc_ex6.py:19]: thanks

File: datastruc_ex8.py (hunk 9 above)
Comment (YuvalRi) [reply datastruc_ex8.py:27]: me too! thanks

File: datastruc_ex7.py (hunk 6 above)
Comment (YuvalRi) [reply datastruc_ex7.py:22]: done.

File: datastruc_ex7.py (hunk 5 above)
Comment (YuvalRi) [reply datastruc_ex7.py:78]: done

File: datastruc_ex7.py (hunk 3 above)
Comment (YuvalRi) [reply datastruc_ex7.py:7]: done.

File: datastruc_ex7.py (hunk 4 above)
Comment (YuvalRi) [reply datastruc_ex7.py:70]: fixed

File: datastruc_ex7.py (hunk 7 above)
Comment (YuvalRi) [reply datastruc_ex7.py:37]: I did not know this tool. done

File: datastruc_ex7.py (hunk 2 above)
Comment (YuvalRi) [reply datastruc_ex7.py:40]: I agree. already changed to more informative one

File: datastruc_ex8.py (hunk 8 above)
Comment (YuvalRi) [reply datastruc_ex8.py:14]: done.

File: datastruc_ex4.py [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+
+# b
+
+# The function 'mergeDict' from the previous task is relevant
+
+
+def taskB(A, B):
+    for key in B:
+        if key in A.keys():
+            lst = [A[key], B[key]]
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex4.py:29]: done

File: datastruc_ex4.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+else:
+    print("There is a problem!")
+
+dic6 = {1: 30, 2: 40}
+dic7 = {2: 20, 1: 90}
+taskB(dic6, dic7)
+if dic6 == {1: [30, 90], 2: [40, 20]}:
+    print("Excellent!")
+else:
+    print("There is a problem!")
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex4.py]: this is a part of task D. fixed in the last commit.

File: datastruc_ex10.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+            all_freq[i] = 1
+    return all_freq
+
+
+def cyclicShift(str1: str, str2: str):
+    '''
+    A function which recieves 2 strings and returns
+    True is str2 is a "cyclic shift" of str1 and False otherwise
+    '''
+    if freqDict(str1) != freqDict(str2):
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex10.py:14]: done by creating option 2

File: datastruc_ex10.py [hunk 18]
```diff
@@ -0,0 +1,36 @@
...
+        return False
+    temp = str1 + str1
+    if (temp.count(str2) > 0):
+        return True
+    else:
+        return False
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex10.py:20]: done by creating option 2

File: datastruc_ex10.py [hunk 19]
```diff
@@ -26,11 +19,62 @@ def cyclicShift(str1: str, str2: str):
...
+    True is str2 is a "cyclic shift" of str1 and False otherwise
+    '''
+    new_str = []
+    for i in range(len(str1)):
+        new_str.append(rotate(str1, i))
+    if str2 in new_str:
+        return True
+    else:
+        return False
+
```
Comment (ofekshmu) [inline datastruc_ex10.py:46]: This implementation in inefficient. I wont accept it. 
1. I dont to check all the possible shifts before returning an answer. I want you to return true upon finding a solution.
2.  There is no need to keep al created rotations in a list.  create a new string -> check it -> decide what to do

File: datastruc_ex10.py [hunk 20]
```diff
@@ -26,11 +19,62 @@ def cyclicShift(str1: str, str2: str):
...
+    '''
+    A function which returns the number of shifts
+    required to reach str2 from str1
+    '''
+    if not cyclicShift2(str1, str2):
+        return -1
+    else:
+        new_str = []
+        for i in range(len(str1)):
+            new_str.append(rotate(str1, i))
```
Comment (ofekshmu) [inline datastruc_ex10.py:60]: Same comment here as before.. Your implementation is inefficient. DONT iterate over all the possible answers when there is a chance you already got it. We did something similar before... see ex8

**Review by ofekshmu — CHANGES_REQUESTED** [review]
Your documentation improved. Implementations are still not good enough.

File: datastruc_ex4.py [hunk 21]
```diff
@@ -1,29 +1,36 @@
...
 
-# The function 'mergeDict' from the previous task is relevant
+
+def mergeDict(dic_a, dic_b):
+    '''
+    A function which merges between two given dictionaries
+    '''
+    for key, value in dic_b.items():
+        dic_a[key] = value
+    return dic_a
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex4.py:24]: thanks

File: datastruc_ex4.py [hunk 22]
```diff
@@ -36,21 +43,25 @@ def taskB(A, B):
...
+    else:
+        print("There is a problem!")
+
+    dic6 = {1: 30, 2: 40}
+    dic7 = {2: 20, 1: 90}
+    taskC(dic6, dic7)
+    if dic6 == {1: [30, 90], 2: [40, 20]}:
+        print("Excellent!")
+    else:
+        print("There is a problem!")
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex4.py]: done

File: datastruc_ex7.py [hunk 23]
```diff
@@ -1,43 +1,32 @@
...
-            for i in range(len(randomlist)):
-                selected_words.append(text_list[randomlist[i]])
-            return selected_words
         else:
-            text_list = allText.split('\n')
-            selected_words = []
-            for i in range(len(lst)):
-                selected_words.append(text_list[lst[i]])
-            return selected_words
+            randomlist = lst
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex7.py:20]: done

File: datastruc_ex7.py [hunk 24]
```diff
@@ -54,26 +43,37 @@ def maxCharacter(dict: dict):
...
+    A function which returns all required values
+    '''
+    dictofreq = freqDict(fromListToString(wordSelected(n=n,
+                                                       random=rand,
+                                                       lst=lst)))
+    max_val = maxCharacter(dictofreq)
+    median_val = findMedian(dictofreq)
+    #print(f'The frequency dictionary is: {dictofreq}\
+    #\nThe most common character is: {max_val}\
+    #\nThe median value is: {median_val}')
```
[the remark this answers is not in the document]
Comment (YuvalRi) [reply datastruc_ex7.py:70]: done

File: datastruc_ex10.py (hunk 19 above)
Comment (YuvalRi) [reply datastruc_ex10.py:46]: done

File: datastruc_ex10.py (hunk 20 above)
Comment (YuvalRi) [reply datastruc_ex10.py]: done

File: datastruc_ex7.py [hunk 25]
```diff
@@ -15,11 +15,11 @@ def wordSelected(n: int, random: bool, lst: list):
...
         if random:
-            randomlist = rnd.sample(range(0, len(text_list)), n)
+            new_lst = rnd.sample(range(0, len(text_list)), n)
         else:
-            randomlist = lst
-        for i in range(len(randomlist)):
-            selected_words.append(text_list[randomlist[i]])
+            new_lst = lst
+        for i in range(len(new_lst)):
+            selected_words.append(text_list[new_lst[i]])
```
Comment (ofekshmu) [inline datastruc_ex7.py:22]: A lot better

File: datastruc_ex7.py [hunk 26]
```diff
@@ -63,15 +63,13 @@ def dataFunction(n: int, rand: bool, lst: list):
...
-        f'''The frequency dictionary is: {dictofreq}
-        The most common character is: {max_val}
-        The median value is: {median_val}''')
+    print(f'The frequency dictionary is: {dictofreq}\
+    \nThe most common character is: {maxCharacter(dictofreq)}\
+    \nThe median value is: {findMedian(dictofreq)}')
+    #print(
+        #f'''The frequency dictionary is: {dictofreq}
+        #The most common character is: {maxCharacter(dictofreq)}
+        #The median value is: {findMedian(dictofreq)}''')
```
Comment (ofekshmu) [inline datastruc_ex7.py:72]: Perfect, See how the first function is too long for the print, so we use a variable, but the other two are compact and short so we can make the code more clean.

File: functionsCheck_ex4.py [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+    else:
+        print("There is a problem!")
+
+    dic6 = {1: 30, 2: 40}
+    dic7 = {2: 20, 1: 90}
+    taskC(dic6, dic7)
+    if dic6 == {1: [30, 90], 2: [40, 20]}:
+        print("Excellent!")
+    else:
+        print("There is a problem!")
```
Comment (ofekshmu) [inline functionsCheck_ex4.py:26]: Creating a file made for testing is a good practice when you have a lot of tests. Keep doing this when your main gets too big or when you are requested to make more than one main. Never use two 'mains' in one file.

**Review by ofekshmu — APPROVED** [review]

**Pull request merged**